File: aiwa_app/android/app/src/main/kotlin/com/example/aiwa_milestone_a/VideoFrameExtractor.kt
package com.example.aiwa_milestone_a

import android.content.Context
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileNotFoundException
import kotlin.math.roundToInt

object VideoFrameExtractor {
    /**
     * Extract frames with progress callback (streaming mode)
     * 
     * @param context Android context
     * @param tokenId Cancellation token ID (empty string if not provided)
     * @param videoPath Absolute path to the video file
     * @param targetFps Desired frames per second for extraction
     * @param maxWidth Maximum width for extracted frames (scaled if larger)
     * @param maxHeight Maximum height for extracted frames (scaled if larger)
     * @param quality JPEG quality (0-100)
     * @param isCancelled Lambda function that returns true if extraction should be cancelled
     * @param onProgress Callback(processed, total, frameData) - frameData is null for progress-only updates
     */
    fun extractFramesWithProgress(
        context: Context,
        tokenId: String = "",
        videoPath: String,
        targetFps: Double,
        maxWidth: Int,
        maxHeight: Int,
        quality: Int,
        isCancelled: () -> Boolean = { false },
        onProgress: (Int, Int, ByteArray?) -> Unit
    ) {
        // Validate file exists
        val videoFile = File(videoPath)
        if (!videoFile.exists()) {
            throw FileNotFoundException("Video file not found: $videoPath")
        }

        android.util.Log.d("VideoFrameExtractor", "Starting extraction: $videoPath")
        val retriever = MediaMetadataRetriever()
        
        try {
            // Initialize retriever
            android.util.Log.d("VideoFrameExtractor", "Setting data source")
            retriever.setDataSource(videoPath)

            // Get video duration
            val durationStr = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
                ?: throw IllegalArgumentException("Cannot get video duration")
            val durationMs = durationStr.toLong()
            val durationUs = durationMs * 1000

            // Calculate frame interval
            val frameIntervalUs = (1_000_000.0 / targetFps).roundToInt().toLong()
            val totalFrames = (durationUs / frameIntervalUs).toInt()

            android.util.Log.d("VideoFrameExtractor", "Video: ${durationMs}ms, Target FPS: $targetFps, Total frames: $totalFrames")

            // Extract frames with progress updates
            for (i in 0 until totalFrames) {
                // Check for cancellation before processing each frame
                if (isCancelled()) {
                    android.util.Log.d("VideoFrameExtractor", "Extraction cancelled at frame $i/$totalFrames")
                    return
                }
                
                val timeUs = i * frameIntervalUs
                
                // Log every 50 frames
                if (i % 50 == 0) {
                    android.util.Log.d("VideoFrameExtractor", "Progress: $i/$totalFrames frames")
                }
                
                // Get frame
                val bitmap = retriever.getFrameAtTime(
                    timeUs,
                    MediaMetadataRetriever.OPTION_CLOSEST
                )

                if (bitmap != null) {
                    // Scale bitmap
                    val scaledBitmap = scaleBitmap(bitmap, maxWidth, maxHeight)
                    
                    // Convert to JPEG
                    val outputStream = ByteArrayOutputStream()
                    scaledBitmap.compress(Bitmap.CompressFormat.JPEG, quality, outputStream)
                    val jpegData = outputStream.toByteArray()
                    
                    // Send frame data + progress
                    onProgress(i + 1, totalFrames, jpegData)
                    
                    // Recycle bitmaps to free memory
                    if (scaledBitmap != bitmap) {
                        scaledBitmap.recycle()
                    }
                    bitmap.recycle()
                } else {
                    // Progress update without frame data
                    onProgress(i + 1, totalFrames, null)
                }
            }

            android.util.Log.d("VideoFrameExtractor", "Extraction complete: $totalFrames frames")

        } catch (e: OutOfMemoryError) {
            android.util.Log.e("VideoFrameExtractor", "Out of memory: ${e.message}")
            throw OutOfMemoryError("Out of memory while extracting frames: ${e.message}")
        } catch (e: IllegalArgumentException) {
            android.util.Log.e("VideoFrameExtractor", "Invalid argument: ${e.message}")
            throw IllegalArgumentException("Invalid video file or arguments: ${e.message}")
        } catch (e: Exception) {
            android.util.Log.e("VideoFrameExtractor", "Extraction failed: ${e.message}")
            throw Exception("Failed to extract frames: ${e.message}", e)
        } finally {
            try {
                retriever.release()
                android.util.Log.d("VideoFrameExtractor", "MediaMetadataRetriever released")
            } catch (e: Exception) {
                // Ignore release errors
            }
        }
    }

    /**
     * Extract frames from a video file using MediaMetadataRetriever (legacy method)
     * 
     * @param context Android context
     * @param videoPath Absolute path to the video file
     * @param targetFps Desired frames per second for extraction
     * @param maxWidth Maximum width for extracted frames (scaled if larger)
     * @param maxHeight Maximum height for extracted frames (scaled if larger)
     * @param quality JPEG quality (0-100)
     * @return List of JPEG byte arrays
     */
    fun extractFrames(
        context: Context,
        videoPath: String,
        targetFps: Double,
        maxWidth: Int,
        maxHeight: Int,
        quality: Int
    ): List<ByteArray> {
        val frames = mutableListOf<ByteArray>()
        
        extractFramesWithProgress(
            context = context,
            tokenId = "", // No token for legacy method
            videoPath = videoPath,
            targetFps = targetFps,
            maxWidth = maxWidth,
            maxHeight = maxHeight,
            quality = quality,
            isCancelled = { false }, // No cancellation for legacy method
            onProgress = { processed, total, frameData ->
                if (frameData != null) {
                    frames.add(frameData)
                }
            }
        )
        
        return frames
    }

    /**
     * Extract frames from a video file using MediaMetadataRetriever (old implementation, kept for reference)
     */
    private fun extractFramesOld(
        context: Context,
        videoPath: String,
        targetFps: Double,
        maxWidth: Int,
        maxHeight: Int,
        quality: Int
    ): List<ByteArray> {
        // Validate file exists
        val videoFile = File(videoPath)
        if (!videoFile.exists()) {
            throw FileNotFoundException("Video file not found: $videoPath")
        }

        val retriever = MediaMetadataRetriever()
        val frames = mutableListOf<ByteArray>()

        try {
            // Initialize retriever with video file
            retriever.setDataSource(videoPath)

            // Get video duration in microseconds
            val durationStr = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
                ?: throw IllegalArgumentException("Cannot get video duration")
            val durationMs = durationStr.toLong()
            val durationUs = durationMs * 1000

            // Calculate frame interval in microseconds
            val frameIntervalUs = (1_000_000.0 / targetFps).roundToInt().toLong()
            
            // Calculate total number of frames
            val totalFrames = (durationUs / frameIntervalUs).toInt()

            // Extract frames at specified intervals
            for (i in 0 until totalFrames) {
                val timeUs = i * frameIntervalUs
                
                // Get frame at specified time
                val bitmap = retriever.getFrameAtTime(
                    timeUs,
                    MediaMetadataRetriever.OPTION_CLOSEST
                )

                if (bitmap != null) {
                    // Scale bitmap if needed
                    val scaledBitmap = scaleBitmap(bitmap, maxWidth, maxHeight)
                    
                    // Convert to JPEG
                    val outputStream = ByteArrayOutputStream()
                    scaledBitmap.compress(Bitmap.CompressFormat.JPEG, quality, outputStream)
                    val jpegData = outputStream.toByteArray()
                    
                    frames.add(jpegData)
                    
                    // Recycle bitmaps to free memory
                    if (scaledBitmap != bitmap) {
                        scaledBitmap.recycle()
                    }
                    bitmap.recycle()
                } else {
                    // If frame extraction fails, add empty placeholder or skip
                    // For now, we skip empty frames
                }
            }

            return frames

        } catch (e: OutOfMemoryError) {
            throw OutOfMemoryError("Out of memory while extracting frames: ${e.message}")
        } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid video file or arguments: ${e.message}")
        } catch (e: Exception) {
            throw Exception("Failed to extract frames: ${e.message}", e)
        } finally {
            // Always release retriever resources
            try {
                retriever.release()
            } catch (e: Exception) {
                // Ignore release errors
            }
        }
    }

    /**
     * Scale a bitmap to fit within max dimensions while maintaining aspect ratio
     */
    private fun scaleBitmap(original: Bitmap, maxWidth: Int, maxHeight: Int): Bitmap {
        val originalWidth = original.width
        val originalHeight = original.height

        // If already smaller than max, return original
        if (originalWidth <= maxWidth && originalHeight <= maxHeight) {
            return original
        }

        // Calculate scaling factor
        val widthScale = maxWidth.toFloat() / originalWidth
        val heightScale = maxHeight.toFloat() / originalHeight
        val scale = minOf(widthScale, heightScale)

        // Calculate new dimensions
        val newWidth = (originalWidth * scale).roundToInt()
        val newHeight = (originalHeight * scale).roundToInt()

        // Create scaled bitmap
        return Bitmap.createScaledBitmap(original, newWidth, newHeight, true)
    }
}

File: aiwa_app/ios/Runner/VideoFrameExtractor.swift
import Foundation
import AVFoundation
import UIKit

/// Custom error type for video frame extraction
struct VideoFrameExtractionError {
    let code: String
    let message: String
    let details: String?
}

/// Video frame extractor using AVFoundation
class VideoFrameExtractor {
    /**
     * Extract frames with progress callback (streaming mode)
     *
     * - Parameters:
     *   - tokenId: Cancellation token ID (empty string if not provided)
     *   - videoPath: Absolute path to the video file
     *   - targetFps: Desired frames per second for extraction
     *   - maxWidth: Maximum width for extracted frames (scaled if larger)
     *   - maxHeight: Maximum height for extracted frames (scaled if larger)
     *   - quality: JPEG quality (0-100)
     *   - isCancelled: Closure that returns true if extraction should be cancelled
     *   - progress: Callback(processed, total, frameData, error) - frameData is nil for progress-only updates
     */
    static func extractFramesWithProgress(
        tokenId: String = "",
        videoPath: String,
        targetFps: Double,
        maxWidth: Int,
        maxHeight: Int,
        quality: Int,
        isCancelled: @escaping () -> Bool = { false },
        progress: @escaping (Int, Int, Data?, VideoFrameExtractionError?) -> Void
    ) {
        // Validate file exists
        let fileURL = URL(fileURLWithPath: videoPath)
        guard FileManager.default.fileExists(atPath: videoPath) else {
            progress(0, 0, nil, VideoFrameExtractionError(
                code: "FILE_NOT_FOUND",
                message: "Video file not found: \(videoPath)",
                details: nil
            ))
            return
        }
        
        print("VideoFrameExtractor: Starting extraction: \(videoPath)")
        
        // Create AVAsset from file URL
        let asset = AVAsset(url: fileURL)
        
        // Get video duration
        let duration = asset.duration
        let durationSeconds = CMTimeGetSeconds(duration)
        
        guard durationSeconds.isFinite && durationSeconds > 0 else {
            progress(0, 0, nil, VideoFrameExtractionError(
                code: "INVALID_VIDEO",
                message: "Cannot get video duration",
                details: nil
            ))
            return
        }
        
        // Create image generator
        let imageGenerator = AVAssetImageGenerator(asset: asset)
        imageGenerator.appliesPreferredTrackTransform = true
        imageGenerator.requestedTimeToleranceBefore = .zero
        imageGenerator.requestedTimeToleranceAfter = .zero
        imageGenerator.maximumSize = CGSize(width: maxWidth, height: maxHeight)
        
        // Calculate frame timestamps
        let frameInterval = 1.0 / targetFps
        let totalFrames = Int(durationSeconds / frameInterval)
        var times: [CMTime] = []
        
        for i in 0..<totalFrames {
            let timeSeconds = Double(i) * frameInterval
            let time = CMTime(seconds: timeSeconds, preferredTimescale: 600)
            times.append(time)
        }
        
        print("VideoFrameExtractor: Video: \(durationSeconds)s, Target FPS: \(targetFps), Total frames: \(totalFrames)")
        
        // Extract frames one by one with progress updates
        var frameIndex = 0
        
        func extractNextFrame() {
            // Check for cancellation before processing
            if isCancelled() {
                print("VideoFrameExtractor: Extraction cancelled at frame \(frameIndex)/\(totalFrames)")
                return
            }
            
            guard frameIndex < times.count else {
                print("VideoFrameExtractor: Extraction complete: \(totalFrames) frames")
                return
            }
            
            let time = times[frameIndex]
            let currentIndex = frameIndex
            
            // Log every 50 frames
            if frameIndex % 50 == 0 {
                print("VideoFrameExtractor: Progress: \(frameIndex)/\(totalFrames) frames")
            }
            
            do {
                let cgImage = try imageGenerator.copyCGImage(at: time, actualTime: nil)
                
                // Convert CGImage to JPEG Data
                if let jpegData = convertToJPEG(cgImage: cgImage, quality: quality) {
                    progress(currentIndex + 1, totalFrames, jpegData, nil)
                } else {
                    progress(currentIndex + 1, totalFrames, nil, nil)
                }
                
                frameIndex += 1
                
                // Check cancellation before continuing
                guard !isCancelled() else {
                    print("VideoFrameExtractor: Extraction cancelled after frame \(frameIndex)/\(totalFrames)")
                    return
                }
                
                // Continue with next frame (with small delay to avoid blocking)
                DispatchQueue.global(qos: .userInitiated).async {
                    extractNextFrame()
                }
                
            } catch {
                print("VideoFrameExtractor: Frame extraction error at \(frameIndex): \(error.localizedDescription)")
                progress(currentIndex + 1, totalFrames, nil, nil)
                frameIndex += 1
                
                // Check cancellation before continuing even on error
                guard !isCancelled() else {
                    print("VideoFrameExtractor: Extraction cancelled after error at frame \(frameIndex)/\(totalFrames)")
                    return
                }
                
                // Continue with next frame even on error
                DispatchQueue.global(qos: .userInitiated).async {
                    extractNextFrame()
                }
            }
        }
        
        // Start extraction
        extractNextFrame()
    }
    
    /**
     * Extract frames from a video file using AVFoundation (legacy method)
     *
     * - Parameters:
     *   - videoPath: Absolute path to the video file
     *   - targetFps: Desired frames per second for extraction
     *   - maxWidth: Maximum width for extracted frames (scaled if larger)
     *   - maxHeight: Maximum height for extracted frames (scaled if larger)
     *   - quality: JPEG quality (0-100)
     *   - completion: Callback with frames array or error
     */
    static func extractFrames(
        videoPath: String,
        targetFps: Double,
        maxWidth: Int,
        maxHeight: Int,
        quality: Int,
        completion: @escaping ([FlutterStandardTypedData]?, VideoFrameExtractionError?) -> Void
    ) {
        var frames: [FlutterStandardTypedData] = []
        
        extractFramesWithProgress(
            videoPath: videoPath,
            targetFps: targetFps,
            maxWidth: maxWidth,
            maxHeight: maxHeight,
            quality: quality
        ) { processed, total, frameData, error in
            if let error = error {
                completion(nil, error)
            } else if let frameData = frameData {
                frames.append(FlutterStandardTypedData(bytes: frameData))
            }
            
            // Complete when done
            if processed >= total {
                completion(frames, nil)
            }
        }
    }
    
    /**
     * Extract frames from a video file using AVFoundation (old async implementation, kept for reference)
     */
    private static func extractFramesOld(
        videoPath: String,
        targetFps: Double,
        maxWidth: Int,
        maxHeight: Int,
        quality: Int,
        completion: @escaping ([FlutterStandardTypedData]?, VideoFrameExtractionError?) -> Void
    ) {
        // Validate file exists
        let fileURL = URL(fileURLWithPath: videoPath)
        guard FileManager.default.fileExists(atPath: videoPath) else {
            completion(nil, VideoFrameExtractionError(
                code: "FILE_NOT_FOUND",
                message: "Video file not found: \(videoPath)",
                details: nil
            ))
            return
        }
        
        // Create AVAsset from file URL
        let asset = AVAsset(url: fileURL)
        
        // Get video duration
        let duration = asset.duration
        let durationSeconds = CMTimeGetSeconds(duration)
        
        guard durationSeconds.isFinite && durationSeconds > 0 else {
            completion(nil, VideoFrameExtractionError(
                code: "INVALID_VIDEO",
                message: "Cannot get video duration",
                details: nil
            ))
            return
        }
        
        // Create image generator
        let imageGenerator = AVAssetImageGenerator(asset: asset)
        imageGenerator.appliesPreferredTrackTransform = true
        imageGenerator.requestedTimeToleranceBefore = .zero
        imageGenerator.requestedTimeToleranceAfter = .zero
        imageGenerator.maximumSize = CGSize(width: maxWidth, height: maxHeight)
        
        // Calculate frame timestamps
        let frameInterval = 1.0 / targetFps
        let totalFrames = Int(durationSeconds / frameInterval)
        var times: [NSValue] = []
        
        for i in 0..<totalFrames {
            let timeSeconds = Double(i) * frameInterval
            let time = CMTime(seconds: timeSeconds, preferredTimescale: 600)
            times.append(NSValue(time: time))
        }
        
        // Extract frames asynchronously
        var extractedFrames: [FlutterStandardTypedData] = []
        let queue = DispatchQueue(label: "com.example.aiwa.frameExtraction", qos: .userInitiated)
        
        queue.async {
            var hasError = false
            let semaphore = DispatchSemaphore(value: 0)
            
            imageGenerator.generateCGImagesAsynchronously(forTimes: times) { _, cgImage, _, result, error in
                defer { semaphore.signal() }
                
                if let error = error {
                    print("Frame extraction error: \(error.localizedDescription)")
                    return
                }
                
                if result == .succeeded, let cgImage = cgImage {
                    // Convert CGImage to JPEG Data
                    if let jpegData = self.convertToJPEG(cgImage: cgImage, quality: quality) {
                        let flutterData = FlutterStandardTypedData(bytes: jpegData)
                        extractedFrames.append(flutterData)
                    }
                } else if result == .failed {
                    hasError = true
                }
            }
            
            // Wait for all frames to complete
            for _ in 0..<times.count {
                semaphore.wait()
            }
            
            // Return results on main thread
            DispatchQueue.main.async {
                if hasError && extractedFrames.isEmpty {
                    completion(nil, VideoFrameExtractionError(
                        code: "EXTRACTION_FAILED",
                        message: "Failed to extract any frames from video",
                        details: nil
                    ))
                } else {
                    completion(extractedFrames, nil)
                }
            }
        }
    }
    
    /**
     * Convert CGImage to JPEG Data
     *
     * - Parameters:
     *   - cgImage: Source CGImage
     *   - quality: JPEG quality (0-100)
     * - Returns: JPEG data or nil if conversion fails
     */
    private static func convertToJPEG(cgImage: CGImage, quality: Int) -> Data? {
        let uiImage = UIImage(cgImage: cgImage)
        let compressionQuality = CGFloat(quality) / 100.0
        return uiImage.jpegData(compressionQuality: compressionQuality)
    }
}

